linesearch.go

```diff
@@ -10,24 +10,28 @@ import (
 	"github.com/gonum/floats"
 )
 
-// LinesearchMethod represents an abstract optimization method in which
-// a function is optimized through successive line search optimizations.
-// It consists of a NextDirectioner, which specifies the search direction
-// of each linesearch, and a Linesearcher which performs a linesearch along
-// the search direction.
+// LinesearchMethod represents an abstract optimization method in which a
+// function is optimized through successive line search optimizations.
 type LinesearchMethod struct {
+	// NextDirectioner specifies the search direction of each linesearch.
 	NextDirectioner NextDirectioner
-	Linesearcher    Linesearcher
+	// Linesearcher performs a linesearch along the search direction.
+	Linesearcher Linesearcher
 
-	x     []float64 // Starting point for the current iteration.
-	dir   []float64 // Search direction for the current iteration.
-	first bool      // Indicator of the first iteration.
+	x   []float64 // Starting point for the current iteration.
+	dir []float64 // Search direction for the current iteration.
 
-	evalType EvaluationType
-	iterType IterationType
+	first     bool // Indicator of the first iteration.
+	nextMajor bool // Indicates that MajorIteration must be requested at the next call to Iterate().
+
+	loc  Location  // Storage for intermediate locations.
```

Comment:
When you get here, @btracey , this is the key point in the discussion about preserving the contents of Location. If we are ok with saying that LinesearchMethod must receive the non-evaluated fields of Location unchanged, then this field can go away and we save (in some cases many) unnecessary copies of Gradient (and Hessian in case of Newton). Instead, we could build the location gradually in the Location passed to Init and Iterate.

For example conjugate gradient produces badly scaled directions and the line searches rarely finish in one iteration. Every point that does not satisfy Armijo or Wolfe means an extra copy. For expensive functions probably not significant, but still ...